operators/pkg/controller/elasticsearch/certificates/transport/reconcile.go

```diff
@@ -52,102 +50,47 @@ func ReconcileTransportCertificateSecrets(
 		additionalCAs = append(additionalCAs, []byte(trustRelationship.Spec.CaCert))
 	}
 
-	// get all existing transport certificate secrets for this cluster
-	certificateSecrets, err := findTransportCertificateSecrets(c, es)
-	if err != nil {
+	var pods corev1.PodList
+	if err := c.List(&client.ListOptions{
+		LabelSelector: label.NewLabelSelectorForElasticsearch(es),
+	}, &pods); err != nil {
 		return reconcile.Result{}, err
 	}
 
-	for _, secret := range certificateSecrets {
-		// retrieve pod associated to this secret
-		podName, ok := secret.Labels[LabelAssociatedPod]
-		if !ok {
-			return reconcile.Result{}, fmt.Errorf("Cannot find pod name in labels of secret %s", secret.Name)
-		}
-
-		var pod corev1.Pod
-		if err := c.Get(types.NamespacedName{Namespace: secret.Namespace, Name: podName}, &pod); err != nil {
-			if !apierrors.IsNotFound(err) {
-				return reconcile.Result{}, err
-			}
-
-			// pod does not exist anymore, garbage-collect the secret
-			// give some leniency in pods showing up only after a while
-			if secret.CreationTimestamp.Add(5 * time.Minute).Before(time.Now()) {
-				// if the secret has existed for too long without an associated pod, it's time to GC it
-				log.Info("Unable to find pod associated with secret, GCing", "secret", secret.Name)
-				if err := c.Delete(&secret); err != nil {
-					return reconcile.Result{}, err
-				}
-			} else {
-				log.Info("Unable to find pod associated with secret, but secret is too young for GC", "secret", secret.Name)
-			}
-			continue
-		}
-
+	for _, pod := range pods.Items {
```

Comment:
I'm wondering what happens if we work with a stale cache here: for example a pod that is actually deleted in the apiserver. We would then re-create its secret even though the pod does not exist? Do we need an extra gc mechanism to prevent that from happening?

Reply:
That is indeed possible, and we have that GC mechanism in place (the `cleanup` package).